packages/form-layout/src/vaadin-form-item.js

```diff
@@ -159,17 +160,62 @@ class FormItem extends ThemableMixin(PolymerElement) {
     return 'vaadin-form-item';
   }
 
+  constructor() {
+    super();
+    this.__updateInvalidState = this.__updateInvalidState.bind(this);
+    this.__contentFieldObserver = new MutationObserver(() => this.__updateRequiredState(this.__contentField.required));
+  }
+
   /** @private */
   _onLabelClick() {
-    const firstContentElementChild = Array.prototype.find.call(
-      this.$.contentSlot.assignedNodes(),
-      (e) => e.nodeType === Node.ELEMENT_NODE
-    );
+    const firstContentElementChild = this.__getContentChildNodes()[0];
     if (firstContentElementChild) {
       firstContentElementChild.focus();
       firstContentElementChild.click();
     }
   }
+
+  /** @private */
+  __getContentChildNodes() {
+    return [...this.$.contentSlot.assignedNodes()].filter((e) => e.nodeType === Node.ELEMENT_NODE);
+  }
+
+  /** @private */
+  __onContentSlotChange() {
+    if (this.__contentField) {
+      // Discard the old field
+      this.__updateRequiredState(false);
+      this.__contentFieldObserver.disconnect();
+      delete this.__contentField;
+    }
+
+    const contentFields = this.__getContentChildNodes().filter((node) => 'checkValidity' in node);
+    if (contentFields.length === 1) {
+      // There's only one child field
+      this.__contentField = contentFields[0];
+      this.__updateRequiredState(this.__contentField.required);
+      this.__contentFieldObserver.observe(this.__contentField, { attributes: true });
```

Comment:
Consider using `attributeFilter` to specify that we only care about `required`:

```suggestion
      this.__contentFieldObserver.observe(this.__contentField, { attributes: true, attributeFilter: ['required'] });
```

We have one example of using it with `dir` attribute:

https://github.com/vaadin/web-components/blob/0cbdfdbe4c93a57f1554e426de93ad5788aa6b4e/packages/component-base/src/dir-mixin.js#L22